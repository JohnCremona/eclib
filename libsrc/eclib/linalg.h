// linalg.h: typedefs for scalar, vector, matrix types
//
// Include this to use any of the vec/mat/svec/smat/subspace/ssubspace classes
//
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////

// #define XSTR(x) STR(x)
// #define STR(x) #x
// #ifdef SCALR_OPTION
// #pragma message "The value of SCALAR_OPTION (before): " XSTR(SCALAR_OPTION)
// #else
// #pragma message "SCALAR_OPTION (before) is undefined"
// #endif

#include "modulus.h"
#include "vector.h"
#include "svector.h"
#include "matrix.h"
#include "smatrix.h"
#include "subspace.h"
#include "ssubspace.h"
#include "smatrix_elim.h"

// SCALAR_OPTION may be set to 1 (int), 2 (long), or 3 (bigint) by the
// user.  This determines whether vec, mat, ... default to vec_i,
// mat_i, ... or vac_l, mat_l, ..., or vec_m, mat_m, ... in user code.
// The default value set here is what these abbreviations mean when
// the user does not specify a different option.  NB The default also
// controls the option with which the library is built, which affects
// the linear algebra scalar type used in the modular symbol code
// (which is not currently templated).

#ifndef SCALAR_OPTION
//#define SCALAR_OPTION 1  // int
//#define SCALAR_OPTION 2  // long
#define SCALAR_OPTION 3  // bigint
#endif

// #pragma message "The value of SCALAR_OPTION (after): " XSTR(SCALAR_OPTION)

#undef scalar_type // since this file may be included more than once

#define PRIME30  1073741789  // = largest p such that p < 2^30.

#if (SCALAR_OPTION==2) // long
// #pragma message "In branch with SCALAR_OPTION 2"
#define scalar_type string("long")
typedef long scalar;
typedef vec_l vec;
typedef mat_l mat;
typedef subspace_l subspace;
typedef ssubspace_l ssubspace;
typedef svec_l svec;
typedef smat_l smat;
typedef smat_l_elim smat_elim;
typedef form_finder_l form_finder;
#define to_vec to_vec_l
#define to_mat to_mat_l
#else
#if (SCALAR_OPTION==3) // bigint
// #pragma message "In branch with SCALAR_OPTION 3"
#define scalar_type string("bigint")
typedef bigint scalar;
typedef vec_m vec;
typedef mat_m mat;
typedef subspace_m subspace;
typedef ssubspace_m ssubspace;
typedef svec_m svec;
typedef smat_m smat;
typedef smat_m_elim smat_elim;
typedef form_finder_m form_finder;
#define to_vec to_vec_m
#define to_mat to_mat_m
#else
#if (SCALAR_OPTION==1) // int
// #pragma message "In branch with SCALAR_OPTION 1"
#define scalar_type string("int")
typedef int scalar;
typedef vec_i vec;
typedef mat_i mat;
typedef subspace_i subspace;
typedef ssubspace_i ssubspace;
typedef svec_i svec;
typedef smat_i smat;
typedef smat_i_elim smat_elim;
typedef form_finder_i form_finder;
#define to_vec to_vec_i
#define to_mat to_mat_i
#else
#pragma message "In branch with SCALAR_OPTION not 1, 2, or 3"
#define scalar_type string("undefined")
#endif
#endif
#endif
