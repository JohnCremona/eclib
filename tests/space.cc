// space.cc: Test of subspace package
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
#include <eclib/linalg.h>

const scalar modulus(default_modulus<scalar>());

int main()
{
cout << "Subspace test program with scalar type " << scalar_type << "\n\n";

int times,ntimes,r=0;

while (cout << "Enter size of square matrix M: ", cin >> r, r>0 )
{
  mat m(r,r);
  cout << "Enter entries of M: ";
  cin >> m;
  cout << " M = \n" << m << endl;
  cout << "Trace(M) = " << m.trace() << endl;
/*
  int i;
  mat mpower=m;
  for (i=2; i<=r; i++)
    {mpower=mpower*m;
     cout << "m^" << i << " = \n" << mpower << endl;
     cout << "Trace(m^" << i << ") = " << mpower.trace() << endl;
   }
  {
    vector<long> cp = m.charpoly();
    cout << "char. poly. of m has coefficients " << cp << endl;
  }
  cout << "det(M) = " << m.determinant() << endl;
  cout << "rank(M) = " << m.rank() << endl;
  cout << "nullity(M) = " << m.nullity() << endl;
*/
  cout << endl << "Enter number of times to repeat kernel tests: ";
  cin >> ntimes;
  {
    subspace ker;
    times=ntimes;
    while(times--) ker = kernel(m);
    mat kerbasis = basis(ker);
    cout << "kernel(m) has basis\n" << kerbasis << endl;
    vec_i kerpivs = pivots(ker);
    cout << "pivots: " << kerpivs << "\n";
    scalar kerdenom = denom(ker);
    cout << "denom:  " << kerdenom  << "\n";
  }
  scalar p = modulus;
  cout << "Now compute kernel mod p, p = " << p << endl;
  {
    times=ntimes; subspace ker;
    while(times--) ker = pkernel(m, p);
    mat kerbasis = basis(ker);
    cout << "kernel(m) has basis\n" << kerbasis << endl;
    vec_i kerpivs = pivots(ker);
    cout << "pivots: " << kerpivs << "\n";
    subspace oldker;
    times=ntimes;
    while(times--) oldker = oldpkernel(m, p);
    if((kerbasis!=basis(oldker))||(kerpivs!=pivots(oldker)))
      {
	cout << "!!! Differs from old version !!!" << endl;
	cout << "Old basis = \n"<<basis(oldker)<<endl;
      }
    else
      cout << "!!! Agrees  with old version !!!" << endl;
  }
/*
  {
    subspace im = image(m);
    cout << "image(m) has basis\n" << basis(im);
    cout << "pivots: " << pivots(im) << "\n";
    cout << "denom:  " << denom(im)  << "\n";
  }
  {
    scalar lambda;
    cout << "Enter lambda: "; cin >> lambda;
    subspace elambda = eigenspace(m,lambda);
    cout << "eigenspace for lambda has basis\n" << basis(elambda);
    cout << "with dimension " << dim(elambda) << endl;
  }
*/
}
 cout<<endl;
}
