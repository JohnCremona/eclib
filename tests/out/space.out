
Subspace package test program

Enter size of square matrix M: Enter entries of M:  M = 
[[1,2,3],
[4,5,6],
[7,8,9]]
Trace(M) = 15

Enter number of times to repeat kernel tests: kernel(m) has basis

[[1],
[-2],
[1]]
pivots: [3]
denom:  1
Now compute kernel mod p, p = 1073741789
kernel(m) has basis

[[1],
[-2],
[1]]
pivots: [3]
!!! Agrees  with old version !!!
Enter size of square matrix M: 
